src/crimson/osd/replicated_backend.cc

```diff
@@ -103,3 +112,1005 @@ void ReplicatedBackend::got_rep_op_reply(const MOSDRepOpReply& reply)
     }
   }
 }
+
+seastar::future<> ReplicatedBackend::recover_object(
+  const hobject_t& soid,
+  eversion_t need)
+{
+  logger().debug("{}: {}, {}", __func__, soid, need);
+  auto& wfor = recovering[soid];
+  return seastar::do_with(PushOp(), get_shards_to_push(soid),
+    [this, soid, need, &wfor](auto& pop, auto& shards) {
+    return [this, soid, need, &wfor] {
+      pg_missing_tracker_t local_missing = pg.get_local_missing();
+      if (local_missing.is_missing(soid)) {
+	PullOp po;
+	auto& pi = wfor.pi;
+	prepare_pull(po, pi, soid, need);
+	auto msg = make_message<MOSDPGPull>();
+	msg->from = pg.get_pg_whoami();
+	msg->set_priority(pg.get_recovery_op_priority());
+	msg->pgid = pg.get_pgid();
+	msg->map_epoch = pg.get_osdmap_epoch();
+	msg->min_epoch = pg.get_last_peering_reset();
+	std::vector<PullOp> pulls;
+	pulls.push_back(po);
+	msg->set_pulls(&pulls);
+	(void)shard_services.send_to_osd(pi.from.osd,
```

Comment:
I think you're dropping the send future here.  I don't think that's a good idea.  We want this task to block until those messages actually go out.

Reply:
Um...I chose to wait for the future of "pull" the resolve of which would indicate the message is actually sent out, does this sound reasonable?

Reply:
You mean you are later on waiting on a future stashed in the pull?  Still, seems easier to just return a future which waits on those sends here.